### ISSUE TITLE
conan-qt/5.14.(1/2) broken for macOS ( glibc / libgio ) due to glib no longer static linked

### ISSUE THREAD

**Issue by EugenMayer** [bincrafters/community#1167]
It seem since https://github.com/bincrafters/conan-qt/tree/a5585a998f863d57a5601b2ab620ba9f8ef06641 the macOS build is broken when glib is enabled

I'am yet not sure it could be related to

- upgrading glib: https://github.com/bincrafters/conan-qt/commit/10c0e0622dac3b5312670c236e6bc54ac682579e
- changing the build tools: https://github.com/bincrafters/conan-qt/commit/293c80423b467ab33942e1e379f529f1577ae32a
- and this: https://github.com/bincrafters/conan-qt/commit/d5b3d6b08b7901318b2d44edc51c775613692a54

But when i build against QT 5.14.1 or QT 5.14.2 on appveyor (mojave) i get:

```
dyld: Library not loaded: /Users/travis/.conan/data/glib/2.64.0/bincrafters/stable/package/de25c69fc047f9def627d6b66b5d09535e19ecf9/lib/libgio-2.0.0.dylib
  Referenced from: /Users/appveyor/.conan/data/qt/5.14.2/kwconan/stable/package/1137bd162044cde6302038a8baedf114cdc34c7c/lib/libQt5Core.5.14.2.dylib
  Reason: image not found
```

building under azure pipelines ( mojava ) it actually compiles ( since i expect there is some glibc installed via brew already ) -but when then running in app on Mojave/Catalina one gets:

```
dyld: launch, loading dependent libraries

Dyld Error Message:
  Library not loaded: /Users/*/libgio-2.0.0.dylib
  Referenced from: /Applications/<REDACTED>/Contents/Frameworks/libQt5Core.5.dylib
  Reason: image not found
```

The old 5.14.1 builds from Feb1 do not have this issue. Any suggestions or assumptions?

my flags:

```

[options]
qca:shared=True
qt:opengl=no
qt:qttools=True
qt:shared=True
qt:with_fontconfig=True
qt:with_harfbuzz=False
qt:with_libalsa=False
qt:with_libjpeg=False
qt:with_libpng=True
qt:with_mysql=False
qt:with_odbc=False
qt:with_openal=False
qt:with_pcre2=True
qt:with_pq=False
qt:with_sdl2=False
qt:with_sqlite3=False
```

if of any use, you find our compile build on bincrafters at:

https://bintray.com/eugenmayer/oss/qt%3Akwconan/5.14.1%3Astable

And this should be the working version: https://bintray.com/eugenmayer/oss/qt%3Akwconan/5.14.1%3Astable#files/kwconan%2Fqt%2F5.14.1%2Fstable%2F0%2Fpackage%2Ff460a8a82ded8a94bb8f94f8777d1b34ee4677b9%2F0

**Comment by ericLemanissier** [comment]
Could you please send us the full log of the build which failed, or the link to the CI run ?

**Comment by EugenMayer** [comment]
@ericLemanissier which build you are refering too, the one we have used to build our QT version or the one of the app linked against?

**Comment by ericLemanissier** [comment]
both please.

**Comment by EugenMayer** [comment]
What we could find out is that the old build:

https://bintray.com/eugenmayer/oss/qt%3Akwconan/5.14.1%3Atesting#files/kwconan%2Fqt%2F5.14.1%2Ftesting%2F0%2Fpackage%2F1137bd162044cde6302038a8baedf114cdc34c7c%2F0

has been using `glib:shared=False`

But all recent builds are using

`glib:shared=True`

https://bintray.com/eugenmayer/oss/qt%3Akwconan/5.14.1%3Astable#files/kwconan%2Fqt%2F5.14.1%2Fstable%2F0%2Fpackage%2F1137bd162044cde6302038a8baedf114cdc34c7c%2F0

Are you aware of this change in maybe glib

**Comment by EugenMayer** [comment]
I guess it all comes down to https://github.com/bincrafters/conan-glib/commit/3b5a2273d8da7b874973fdd7ea4af3e105c343b1#diff-2b1d42f71f22b7ea0412d7602dec166f .. are there any reasons you did so?

**Comment by ericLemanissier** [comment]
yes, having a default value inconsistent with the upstream library leads to all sorts of breakage in other packages. Please send the log if you want us to help you further.
I just downloaded https://bintray.com/bincrafters/public-conan/download_file?file_path=bincrafters%2Fglib%2F2.64.0%2Fstable%2F0%2Fpackage%2Fde25c69fc047f9def627d6b66b5d09535e19ecf9%2F0%2Fconan_package.tgz and it contains the file `libgio-2.0.0.dylib` in the lib folder, so I have no clue why it is not found in the two error messages you pasted in your first message.

**Comment by EugenMayer** [comment]
@ericLemanissier i think the main issue here is, that you do no have a "package maintainer version" to the actual QT version. I understand, you are packagin 5.14.1 or 5.14.2 - but your are constanly changing the recipe in the stable/<> branch.

And you do things like this, which then affects people down the pipeline. Having something like 5.14.1-2 where the `2` stands for the package version would really help to understand why things break or to trace - but we have a "morphing" package all the time.

To your question, we are packaging QT in our app dynamically and yet, we did not need to include `libgio-2.0.0.dylib` (since it was statically build). This is what you changed with the upgrade of glib in 5.14.1 after the initial release of you 5.14.1 and we simply did not know about it. There are no changelogs, nothing, one has to basically trace it back through the commits on the `conan-qt` repo, then go down the path for every requirement or simply try to diff the `conanfile.txt` or whatever.

Fairly heavy lifting for that sort of issue. 

What do you suggest people should do when building on top of the QT conan-recpie using own flags and being dependent on somewhat consistent builds at least for a static patch version like 5.14.1? Should we fork it ourselfs and do upstream merges?

Beside that, would you just pick "all libs in the libs" folder and ship them "with the app" and not pick them by hand? What we do right now is

CMakeList.txt
```
# Include external libraries
# -> Qt5
SET(QT_MIN_VERSION "5.14.0")
find_package(Qt5Core ${QT_MIN_VERSION})
set_package_properties(Qt5Core PROPERTIES
    DESCRIPTION "Core library of the Qt5 framework"
    URL "http://www.qt.io"
    TYPE REQUIRED
    PURPOSE "Core functionality, and also file watch (QFileSystemWatcher), JSON parsing (QJsonDocument), command line parsing (QCommandLineParser)."
)
find_package(Qt5Gui ${QT_MIN_VERSION})
set_package_properties(Qt5Gui PROPERTIES
    DESCRIPTION "Graphical user interface library of the Qt5 framework"
    URL "http://doc.qt.io/qt-5/qtgui-index.html"
    TYPE REQUIRED
    PURPOSE "GUI (tray icon) and opening files/folders with default application (QDesktopServices)."
)
find_package(Qt5Network ${QT_MIN_VERSION})
set_package_properties(Qt5Network PROPERTIES
    DESCRIPTION "Network library of the Qt5 framework"
    URL "http://doc.qt.io/qt-5/qtnetwork-module.html"
    TYPE REQUIRED
    PURPOSE "Network communication with the platform, including HTTP, proxies, local socket communication (QNetworkAccessManager, QLocalServer/Socket)."
)
find_package(Qt5Widgets ${QT_MIN_VERSION})
set_package_properties(Qt5Widgets PROPERTIES
    DESCRIPTION "Widgets library of the Qt5 framework"
    URL "http://doc.qt.io/qt-5/qtwidgets-module.html"
    TYPE REQUIRED
    PURPOSE "GUI dialogs (QMessageBox) for the user."
)
find_package(Qt5LinguistTools ${QT_MIN_VERSION})
set_package_properties(Qt5LinguistTools PROPERTIES
    DESCRIPTION "Linguist/Internationalization tools for the Qt5 framework"
    URL "http://doc.qt.io/qt-5/qtlinguist-index.html"
    TYPE REQUIRED
    PURPOSE "Create/update translation files (lupdate, lrelease, QObject::tr)."
)
find_package(Qt5Test ${QT_MIN_VERSION})
set_package_properties(Qt5Test PROPERTIES
    DESCRIPTION "Testing library of the Qt5 framework"
    URL "http://doc.qt.io/qt-5/qttest-module.html"
    TYPE REQUIRED
    PURPOSE "Testing framework, for testing of GoEdit only (QTest, QSignalSpy)."
)
find_package(Qt5Xml ${QT_MIN_VERSION})
set_package_properties(Qt5Xml PROPERTIES
    DESCRIPTION "XML library of the Qt5 framework"
    URL "http://doc.qt.io/qt-5/qtxml-module.html"
    TYPE REQUIRED
    PURPOSE "Parsing for HTML documents, for testing only (QDomDocument etc.)."
)
if(APPLE)
find_package(Qt5DBus ${QT_MIN_VERSION})
set_package_properties(Qt5DBus PROPERTIES
    DESCRIPTION "DBus library of the Qt5 framework"
    URL "http://doc.qt.io/qt-5/qtdbus-index.html"
    TYPE REQUIRED
    PURPOSE "Required by cocoa platform plugin."
)

find_package(Qt5PrintSupport ${QT_MIN_VERSION})
set_package_properties(Qt5PrintSupport PROPERTIES
    DESCRIPTION "Portable printing support library of the Qt5 framework"
    URL "https://doc.qt.io/qt-5/qtprintsupport-index.html"
    TYPE REQUIRED
    PURPOSE "Required by cocoa platform plugin.."
)
endif(APPLE)
```

Would you somewhat glob them so in case that a dependency no is dynamically linked it is not missed?

Thanks!

**Comment by ericLemanissier** [comment]
What you describe regarding package version is true, but It is simply not possible to track individual changes at the package level, because it does not scale, especially with packages with as much dependencies of qt.
The best way to tackle this is to create a bintray remote (you seem to have one already), where you upload packages known to work. then you do not depend on bincrafters and even conan-center-index. You probably won't have to fork the recipes. If at some point you feel the need to migrate to a newer version of a recipe/library, then it's up to you to do the tests/inspection of the changes of the recipe, and once it is done you can upload it to your remote.
If you don't want to go through this process, then your best shot is to enforce the assumption you make about recipe, eg assumption:"I don't deploy library X because it is static" enforcement:"set option X:shared=False", but it's true that it is often hard to explicit all of one's assumptions.
Last option: be extremely agile, and live at head for everything, but that requires constant testing, and being able to fix whatever happens quickly.
Regarding deployment, I have no experience on mac, but there is a lot of information on https://doc.qt.io/qt-5/macos-deployment.html

**Comment by EugenMayer** [comment]
Thank you, i have to look at those options.

running into a new issue though, it seems to work for macOS and Linux now, but breaks for Windows if i use the `glib:shared=False` flag there due to 

```
ERROR: glib/2.64.0@bincrafters/stable: qt/5.14.2@kwconan/stable tried to change glib/2.64.0@bincrafters/stable option shared to False
but it was already assigned to False by qt/5.14.2@kwconan/stable
```

Do i really assume right that conan does treat this as an error since the option is already what was expected, thus it errors out?

**Comment by ericLemanissier** [comment]
It is because glib does not support static package on windows, so it forces shared build : https://github.com/bincrafters/conan-glib/blob/testing/2.64.0/conanfile.py#L42 . It would be cleaner to raise a ConanInvalidConfiguration if shared is False on windows, now that the default is shared=True. As a workaround you could use option qt:with_glib=False.

**Comment by EugenMayer** [comment]
Interesting, do you mean `qt:with_glib=False` for windows only?

**Comment by EugenMayer** [comment]
Changed the title for people finding this over google and maybe being helpful to them. Be free to adjust to whatever you think is best

**Comment by ericLemanissier** [comment]
For your information, [glibc ](https://www.gnu.org/software/libc/)is a complete different thing. The problem here is with [glib](https://developer.gnome.org/glib/).
You should disable with_glib on all platforms on which glib is used only by qt. After all the trouble I have been through with glib, I wonder if the default with_glib should not be False...
Regarding the first error message you copied, it mentions `/Users/travis/.conan/` which is a path used during build of glib itself. It must be hard-coded somewhere during build, and it does not find it when consuming glib on appveyor. I don't understand where does this absolute path come from, it does not appear anywhere else in the log https://ci.appveyor.com/project/bincrafters/conan-qt/builds/32010740?fullLog=true

**Comment by vkhaitan** [comment]
I just wanted to point towards the real issue --

see the conan-qt binary build here -> https://bintray.com/bincrafters/public-conan/download_file?file_path=bincrafters%2Fqt%2F5.14.2%2Fstable%2F0%2Fpackage%2Fb5bcc198f5c7778c724c7944c7c1df2931ad085e%2F0%2Fconan_package.tgz

then use => otool -L libQt5Core_debug.5.14.2.dylib
Ther result ===>  /Users/travis/.conan/data/glib/2.64.0/bincrafters/stable/package/f751815c7759075d66608c47ddf9531a9ba2154a/lib/libglib-2.0.0.dylib (compatibility version 6401.0.0, current version 6401.0.0)

That solve the mystery of "travis" user. Basically, for now, your mac build is broken. That needs to be fixed.

**Comment by ericLemanissier** [comment]
thanks for confirming what I detected in my last message. I must be missing something because I fail to see how that solves any mystery though. Another mystery is how we manage to build qt on azure pipelines without errors (no travis user either there)

**Comment by vkhaitan** [comment]
this conan_package was built on machine where travis user exist. Now this libQt5Core dependency is an issue during runtime, not compile time, so you can build it fine. Not just that, you can build any Qt software fine too.
The problem is that the Qt software wont start and complain about the non-existent dependency.

**Comment by vkhaitan** [comment]
No, the path is not hardcoded. It is normal to have this dependency initially. After the library build, this path is change to relative path => @rpath/libdlib-2.0.0.dylib  . So, the issue is that we need to add code to do that in conan build.

**Comment by ericLemanissier** [comment]
https://ci.appveyor.com/project/bincrafters/conan-qt/builds/32010740?fullLog=true#L2537 shows a build failure of qt itself, because of the missing dependency. I don't know why you have your local conan cache in /Users/travis/.conan, but this does not seem like a reasonable thing to impose to conan users.

**Comment by vkhaitan** [comment]
the travis mystery  probably comes from 
https://travis-ci.org/
glib probably is being built on travis CI . There is no other source of this travis user.
It means that during dynamic linking of Qt library, it is using the glib prepared on that CI and hard-coding the path inside the glib linking file. So, it is really and issue of glib then.

**Comment by ericLemanissier** [comment]
Ok, so we definitely agree on the cause (this is exactly what I was trying to describe in my poorly written [message](https://github.com/bincrafters/community/issues/1167#issuecomment-610412147)) 
My current theory is that in all these cases, the glib/gio library is not found because the DYLD_LIBRARY_PATH is not correctly set.
During qt build, conan sets it at the top level with this [`run_environment=True`](https://github.com/bincrafters/conan-qt/blob/testing/5.14.2/conanfile.py#L669) argument, [but macos does not propagate DYLD_LIBRARY_PATH to the subprocesses](https://stackoverflow.com/questions/35568122/why-isnt-dyld-library-path-being-propagated-here/35570229#35570229).  The build failure happens in `qvkgen_wrapper.sh` call, which cannot know the DYLD_LIBRARY_PATH because macos did not propagate it. I'll see if there is a way to make qt explicitly set DYLD_LIBRARY_PATH for subprocesses.
Could you try to solve you run-time problem by setting DYLD_LIBRARY_PATH  to point to the folder containing the missing lib ?

**Comment by ericLemanissier** [comment]
I can confirm that the bug you faced to begin with is actually https://bugreports.qt.io/browse/QTBUG-57204. I applied the suggested fix and the build actually works now: https://ci.appveyor.com/project/bincrafters/conan-qt/builds/32026251 (even when some dependencies are not provided by the system)
The fix is now available on the 5.14.2/testing branch. @EugenMayer  could you test it ?

**Comment by EugenMayer** [comment]
Thank you for all the effort @ericLemanissier - my problem is, i have no easy way to test this. 

I would need to change my profiles again, then rebuild QT for all platforms, then rebuild QCA against this QT and then try to build my App against this new version. 

Then i would need to revert it, build QT again with shared=false as i desire for production and rebuild QCA and my app again.

AFAIU you could you pick the build of the process you just started on appveyor and the inspect qtcore.dylib with otool and see how glib is looked for - that shoudl be already sufficient for this test?

**Comment by yuriy-mochurad** [comment]
I have just faced this issue with latest qt/5.15.0 stable. Was this fix actually tested?

**Comment by ericLemanissier** [comment]
it was tested on https://ci.appveyor.com/project/bincrafters/conan-qt/builds/32026251?fullLog=true because I don't have access to a mac. Could you please report an issue on https://github.com/bincrafters/community/issues/new?labels=bug&template=package_bug.md&title=%3CLIBRARY-NAME%3E%2F%3CLIBRARY-VERSION%3E%3A+SHORT+DESCRIPTION and give as much detail as possible (especially the full log of the build). Thanks

**Comment by bilke** [comment]
I tested 5.14.2/testing and it seems to work.

**Comment by ericLemanissier** [comment]
thanks for the feedback !

**Comment by bilke** [comment]
Sorry, 5.14.2/testing **does not** work, the fix (https://github.com/bincrafters/conan-qt/commit/91f7f5b5abe7c9201a45bde463fdd57b4d7ff422) is in stable/5.12.8 but not in 5.14.2.

**Comment by ericLemanissier** [comment]
that's surprising, I can see in [forward DYLD_LIBRARY_PATH during build](https://github.com/bincrafters/conan-qt/commit/bd06394e8d2a6e04fc7ce4de915530986f8a5091) https://github.com/bincrafters/conan-qt/commits/stable/5.14.2
Please provide full build log
